docs/client-design.md

```diff
@@ -0,0 +1,184 @@
+
+#### Request Cancelling
+
+The application must be able to cancel a Tntcxx Client request.
+
+#### Error Handling
+
+There must be a convenient exception-safe way for the application to handle 
+errors arising throughout the Tntcxx Client request lifecycle. A request error 
+returned either through the callback, or through the future. 
+
+#### Request Handling
+
+In order for the application to be able to manage a request, a request handle is 
+always returned. The application can check the request status, cancel the 
+request, handle request errors and retrieve the response through this handle 
+(only once). However, if the response was retrieved by other means (either 
+returned through a callback or collected through scatter-gather), the handle
+cannot return the response.
+
+#### Request Timeout
+
+Since the Tntcxx Client does not have control over the application's event 
+processing loop, the application must implement its own request timeouts. The 
+application can cancel stale requests. 
+
+#### Response Lifetime
+
+The response lifetime is managed implicitly and does not require any 
+interference from the application. The response can be retrieved only once, and 
+it is uniquely owned by the application (i.e., it is not copyable). 
+
+#### Request Retrying
+
+Since the Tntcxx Client does not have control over the application's event 
+processing loop, the application must implement its own request retrying.
+
+#### Reconnection
+
+The Tntcxx Client must support implicit reconnection with the same session 
+settings it was created with.
+
+#### Connection Pool
+
+TBD.
+
+#### Transactions
+
+TBD.
+
+#### Failover
+
+TBD.
+
+## Design
+
+### Socket
+
+```c++
+template<class Resource>
+class Socket {
+public:
+	Resource get_resource();
+};
+```
+
+### Network Provider
+
+```c++
+template<class Socket, class Data *>
+class NetProvider {
+public:
+	template<class Socket, class Data *>
+    using read_ready_cb_f = void (*)(Socket sock, Data *data);
+    
+    template<class Socket, class Data *>
+    using write_ready_cb_f = void (*)(Socket sock, Data *data);
+	
+	int register(Socket sock, Data *data, read_ready_cb_f read_ready_cb, write_ready_cb_f write_ready_cb);
+	int unregister(Socket sock);
+};
+```
+
+```c++
+template<class Socket, class Data *>
+class EpollNetProviderData {
+public:
+    EpollNetProviderData(Socket sock, Data *data, read_ready_cb_f read_ready_cb, write_ready_cb_f write_ready_cb);
+    
+    int type;
+    
+    void read_ready();
+    void write_ready();
+};
+```
+
+### Connection
+
+```c++
+struct ConnectionOptions {
+	/* All existing options. */
+	
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+
+template<class NetProvider>
+class Connection {
+public:
+	using request_cb_f = void (*)(std::optional<RequestError> error, Response response, Connection &connection);
```

Comment:
We should note that `RequestError` is not an error sent by Tarantool.

Reply:
Fixed.